TEP-0118: Implement Fanning Out logic to support Matrix Include Parameters in a Task Run (#6341)
tektoncd/pipeline: docs/matrix.md

Reviewer: The docs updates looks great @EmmaMunley, thanks for adding all the examples!

> ## <!--

Reviewer 2: is it possible to remove formatting changes from this PR?

```

### Include

> 

Reviewer 2: Is this feature still in "preview mode"? TBH I'm not sure what "preview mode" means-- can we describe it as an alpha feature instead, and add to the alpha features table?

Author: I think we can remove preview mode now.

> :warning: This feature is in a preview mode.

Explicit combinations of parameters can be specified by `Matrix.Include`.

> The `Matrix.Include` takes `Parameters` of type `"string"` only. For example:

Reviewer 2: can you add the combinations that will be generated from this `include`?

> > It is still in a very early stage of development and is not yet fully functional.

Reviewer 2: can this block be removed?

The `Include` section in the `Matrix` field exists, but is not yet functional.

> The `Include` section in the `Matrix` can be specified with or without `Params` section adding a specific combination of input values for `Matrix Parameters` or defining explicit combinations in the `Matrix` without `Params`.

Reviewer 2: ```suggestion
Explicit combinations of parameters can be specified by `Matrix.Include`.
```

> The `Matrix.Include` will take `Parameters` of type `"string"` only.

Reviewer 2: ```suggestion
The `Matrix.Include` takes `Parameters` of type `"string"` only. For example:
```

  ...
```

#### Include specific combinations in the Matrix

> In the example below, `Pipeline` will need to define a `Matrix` that will execute golang-test in three different linux flavors.

Reviewer 2: ```suggestion
In the example below, the `Pipeline` defines a `Matrix` that will execute golang-test in three different linux flavors.
```

While running on amd64 and ppc64le, golang-test should execute `go test github.com/tektoncd/pipeline/pkg/... -race -cover -v`.
While running on linux/s390x, golang-test should execute `go test github.com/tektoncd/pipeline/pkg/... -cover -v`.

```yaml
apiVersion: tekton.dev/v1beta1
kind: Pipeline
metadata:
  name: pipeline-to-build-and-test-go-project
spec:
  workspaces:
    - name: shared-workspace
  tasks:
    - ...
    - name: golang-test
      taskRef:
        name: golang-test
      workspaces:
        - name: source

>           workspace: shared-workspace

Reviewer 2: suggest removing workspaces in this example (and the next), I don't think it's relevant

      params:
        - name: package
          value: "github.com/tektoncd/pipeline"
        - name: packages
          value: "./pkg/..."
      matrix:
        params:
          - name: GOARCH
            value:
              - "linux/amd64"
              - "linux/ppc64le"
              - "linux/s390x"

>         include:

Reviewer 2: TEP-0118 spells out a really clear example that covers a lot of different cases. We don't necessarily have to put all those cases in one example, but all the cases should be covered in the docs.

Author: Added the additional use cases.

          - name: s390x-no-race
            params:
              - name: GOARCH
                value: "linux/s390x"
              - name: flags
                value: "-cover -v"
    - ...
```

> The above `Matrix` specification will create three `TaskRuns`, one for each architecture in GOARCH ("linux/amd64", "linux/ppc64le", and "linux/s390x") and with default flags. Now, when a `TaskRun` is created for the GOARCH value of linux/s390x, an additional `Parameter` called flags with the value of `-cover -v` will be included in the taskRun.

Reviewer 2: I think it would be helpful to explicitly list out the generated combinations in bullets

Author: Added the generated combinations to the examples

          value: "github.com/tektoncd/pipeline"
        - name: packages
          value: "./pkg/..."
      matrix:
        params:
          - name: GOARCH
            value:
              - "linux/amd64"
              - "linux/ppc64le"
              - "linux/s390x"
        include:
          - name: s390x-no-race
            params:
              - name: GOARCH
                value: "linux/s390x"
              - name: flags
                value: "-cover -v"
    - ...
```

The above `Matrix` specification will create three `TaskRuns`, one for each architecture in GOARCH ("linux/amd64", "linux/ppc64le", and "linux/s390x") and with default flags. Now, when a `TaskRun` is created for the GOARCH value of linux/s390x, an additional `Parameter` called flags with the value of `-cover -v` will be included in the taskRun.

```json!
{ "GOARCH": "linux/amd64", "version": "go1.17" }

{ "GOARCH": "linux/amd64", "version": "go1.18.1" }

{ "GOARCH": "linux/ppc64le", "version": "go1.17" }

{ "GOARCH": "linux/ppc64le", "version": "go1.18.1" }

{ "GOARCH": "linux/s390x", "version": "go1.17", "flags": "-cover -v" }

{ "GOARCH": "linux/s390x", "version": "go1.18.1", "flags": "-cover -v" }

```

> The `Matrix.Include` section can list params which does not exist in the matrix.params section. In this example specification, flags was not listed in `Matrix.Params`. At the same time, `Matrix.Include` section can list params which does exist in the `Matrix.Params` section.

Reviewer 2: This line is a bit confusing/surprising to see in the middle of an example. What I'd do instead is in the section where `include` is first introduced, say something like:
"Each element of matrix.include is a combination of string parameters to use when creating a TaskRun. Paramers in `matrix.include` don't need to be specified in `matrix.params`; see [example] for an example. If parameters in `matrix.include` are also listed in matrix.params, the combinations used for fanned out TaskRuns depend on whether the values of these parameters are the same in matrix.include and matrix.params. see [example] for more details"

        - "linux/s390x"
    - name: version
      value:
        - "go1.17"
        - "go1.18.1"
```

This `matrix` specification will result in six `taskRuns` with the following `matrix` combinations:

```json!
{ "GOARCH": "linux/amd64", "version": "go1.17" }

{ "GOARCH": "linux/amd64", "version": "go1.18.1" }

{ "GOARCH": "linux/ppc64le", "version": "go1.17" }

{ "GOARCH": "linux/ppc64le", "version": "go1.18.1" }

{ "GOARCH": "linux/s390x", "version": "go1.17" }

{ "GOARCH": "linux/s390x", "version": "go1.18.1" }
```

Now, let's introduce `include` with a couple of `Parameters`: `"package"`, `"flags"` and `"context"`:

```yaml=
      matrix:
        params:
          - name: GOARCH
            value:
              - "linux/amd64"
              - "linux/ppc64le"
              - "linux/s390x"
          - name: version
            value:
              - "go1.17"
              - "go1.18.1"
        include:
          - name: common-package
            params:
              - name: package
                value: "path/to/common/package/"
          - name: s390x-no-race
            params:
              - name: GOARCH
                value: "linux/s390x"
              - name: flags

>                 value: "-cover -v"

Reviewer 2: This is redundant with the previous example; do we need both?

```

The fourth `include` clause cannot be added to any original `matrix` combination without overwriting any `params` of the
original combinations, so it is added as an additional `matrix` combination:

```json!
* { **"GOARCH": "I-do-not-exist"** }
```

The above specification will result in seven `taskRuns` with the following matrix combinations:

```json!
{ "GOARCH": "linux/amd64", "version": "go1.17", "package": "path/to/common/package/", "context": "path/to/go117/context" }

{ "GOARCH": "linux/amd64", "version": "go1.18.1", "package": "path/to/common/package/" }

{ "GOARCH": "linux/ppc64le", "version": "go1.17", "package": "path/to/common/package/", "context": "path/to/go117/context" }

{ "GOARCH": "linux/ppc64le", "version": "go1.18.1", "package": "path/to/common/package/" }

{ "GOARCH": "linux/s390x", "version": "go1.17", "package": "path/to/common/package/", "flags": "-cover -v", "context": "path/to/go117/context" }

{ "GOARCH": "linux/s390x", "version": "go1.18.1", "package": "path/to/common/package/", "flags": "-cover -v" }

{ "GOARCH": "I-do-not-exist" }
```

### Define explicit combinations in the Matrix

> In the example below, the user needs to specify explicit mapping between IMAGE and DOCKERFILE, such as:

Reviewer 2: This example is much simpler than the previous, consider putting it first.

```yaml
tasks:

> ---

Reviewer 2: this line is used to separate multiple objects in one yaml file; I think it's a bit confusing to have within the example pipeline spec.
